### PR TITLE
Added Vulkan build steps to makefile

### LINKED ISSUE #17: CICD won't build macOS due to SDK download no longer being available
Currently, the entire project depends on the Vulkan SDK to be available to build. This creates a dependency that can't easily be replicated in a CICD environment. 

At the moment, the CICD system will manually download the SDK to the VM and run the tests from there. This works for Windows and Linux, however MacOS breaks due to a change in the format the SDK is provided in. As such, a new solutions needs to be found to make this work correctly. 

This issue is for changing the build system to pull in the Vulkan dependencies and manually build them. This has the following advantages:
a) Allows the renderer to be build and tested in CICD.
b) Allows Vulkan to be shipped without the end user requiring the Vulkan SDK.

### LINKED ISSUE #20: Optimize DEBUG build system for MacOS
Currently, the DEBUG build process for MacOS is quite slow as it requires duplicate repositories to be pulled and built. There needs to be a way to only build the dependencies once and link them all together.

### PR DESCRIPTION
**Pull request by Raelr**
This PR encapsulates all the revisions to the build system. This PR aims to create a set of steps for creating the necessary Vulkan SDK components needed by the renderer. 

Related issues: https://github.com/Raelr/vk-sandbox/issues/20 https://github.com/Raelr/vk-sandbox/issues/17

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Environment variables for Vulkan. These are all optional.
+
+# VK_LAYER_PATH=$(VULKAN_SDK)/<PATH_TO_LAYERS>
+# VK_ICD_FILENAMES=$(VULKAN_SDK)/<PATH_TO_ICD>
+# DYLD_LIBRARY_PATH=$(VULKAN_SDK)/<PATH_TO_LIBS>
+# VULKAN_INCLUDE_DIR=$(VULKAN_SDK)/<PATH_TO_HEADERS>
```
Comment (jonjondev) [inline .env.template]: These lines should be uncommented and drop the angular brackets; we'll eventually make it so this is an autofill template file.

File: scripts/macos/install.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    mkdirp ${SUBMODULE_INCLUDE_DIR}/vulkan
+
+    cp ${SUBMODULE_DIR}/MoltenVK/External/Vulkan-Headers/include/vulkan/**.h ${SUBMODULE_INCLUDE_DIR}/vulkan 
+    cp ${SUBMODULE_DIR}/MoltenVK/Package/Latest/MoltenVK/dylib/macOS/** ${SUBMODULE_LIB_DIR}/icd.d 
+    cp ${SUBMODULE_DIR}/MoltenVK/Package/Latest/MoltenVK/dylib/macOS/**.dylib ${SUBMODULE_LIB_DIR} 
+
+    cd ${ROOT_DIR}
+}
+
+setup_volk
```
Comment (jonjondev) [inline scripts/macos/install.sh]: this file can be merged with linux install script, just look for uname

File: src/main.cpp [hunk 3]
```diff
@@ -1,3 +1,5 @@
+#define VOLK_IMPLEMENTATION
```
Comment (jonjondev) [inline src/main.cpp:1]: defines usually go below includes

File: scripts/mkdir.bat [hunk 4]
```diff
@@ -0,0 +1,4 @@
+@echo off
+setlocal enableextensions
+md %1
```
Comment (jonjondev) [inline scripts/mkdir.bat:3]: unfortunate that we need this. is this really only for Github Actions?

File: .github/workflows/ubuntu.yml [hunk 5]
```diff
@@ -2,72 +2,40 @@ name: Ubuntu
...
-      
-      - name: update apt 
-        run: sudo apt-get update
-        
+
+      - name: update apt
+        run: |
+          sudo apt-get update
+          sudo apt-get install git build-essential libx11-xcb-dev \
+          libxkbcommon-dev libwayland-dev libxrandr-dev
```
Comment (jonjondev) [inline .github/workflows/ubuntu.yml:22]: what are `libxkbcommon-dev`, `libwayland-dev`, and `libxrandr-dev`?

File: .github/workflows/windows.yml [hunk 6]
```diff
@@ -2,55 +2,38 @@ name: Windows
...
-           $installer = Start-Process -FilePath VulkanSDK.exe -Wait -PassThru -ArgumentList @("/S");
-           $installer.WaitForExit();
-     
+
         # Setup dependencies
       - name: Setup dependencies
-        run: mingw32-make setup
-        
+        run: |
+          ./scripts/install.bat
```
Comment (jonjondev) [inline .github/workflows/windows.yml:25]: this Is all now much cleaner

File: scripts/install.sh [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+setup_volk() {
+    echo "Cloning Volk..."
+    update_submodules volk
+
+    mkdirp ${SUBMODULE_DIR}/include/volk
+
+    cp ${SUBMODULE_DIR}/volk/volk.h ${SUBMODULE_INCLUDE_DIR}/volk 
+    cp ${SUBMODULE_DIR}/volk/volk.c ${SUBMODULE_INCLUDE_DIR}/volk
```
Comment (jonjondev) [inline scripts/install.sh:38]: this is not a header?

File: scripts/install.sh [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    OS="Linux"
+    NUMBER_OF_PROCESSORS="$(nproc)"
+fi
+
+GENERATOR="Unix Makefiles"
+
+SUBMODULE_DIR="${ROOT_DIR}/vendor"
+SUBMODULE_INCLUDE_DIR="${SUBMODULE_DIR}/include"
+SUBMODULE_LIB_DIR="${ROOT_DIR}/lib/${OS}"
```
Comment (jonjondev) [inline scripts/install.sh:19]: might be a useful addition to the file structure

File: scripts/install.sh [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    cp ${SUBMODULE_DIR}/glfw/src/libglfw3.a ${SUBMODULE_LIB_DIR}
+
+    cd ${ROOT_DIR}
+}
+
+setup_moltenVk() {
+    echo "Cloning MoltenVk..."
+
+    update_submodules MoltenVK
+    cd ${SUBMODULE_DIR}/MoltenVK ; ./fetchDependencies --macos --v-headers-root ${SUBMODULE_DIR}/Vulkan-Headers
```
Comment (jonjondev) [inline scripts/install.sh]: can we not just run:
```suggestion
    ./${SUBMODULE_DIR}/MoltenVK/fetchDependencies --macos --v-headers-root ${SUBMODULE_DIR}/Vulkan-Headers
```

File: scripts/install.sh [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+    update_submodules glfw
+
+    cd ${SUBMODULE_DIR}/glfw ; cmake -G "${GENERATOR}" . ; make -j${NUMBER_OF_PROCESSORS}
+
+    mkdirp ${SUBMODULE_LIB_DIR}
+
+    cp ${SUBMODULE_DIR}/glfw/src/libglfw3.a ${SUBMODULE_LIB_DIR}
+
+    cd ${ROOT_DIR}
```
Comment (jonjondev) [inline scripts/install.sh]: if we can avoid uses of `cd`, a failed script won't dump us somewhere we weren't expecting

File: scripts/install.sh [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+setup_vulkan_headers() {
+    update_submodules Vulkan-Headers
+
+    cd ${SUBMODULE_DIR}/Vulkan-Headers ; git fetch --all --tags ; git checkout tags/v1.3.211
+    
+    mkdirp ${SUBMODULE_DIR}/Vulkan-Headers/build
+
+    cd ${SUBMODULE_DIR}/Vulkan-Headers/build ; cmake -DCMAKE_INSTALL_PREFIX=install -G "${GENERATOR}" ..
+    cd ${SUBMODULE_DIR}/Vulkan-Headers/build ; cmake --build . --target install
+
```
Comment (jonjondev) [inline scripts/install.sh]: possible to run `git`/`cmake` on a directory?

File: scripts/install.sh [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+            echo "Setting up Validation Layers..."
+
+            setup_vulkan_validation_layers
+
+            echo "VK_LAYER_PATH=${SUBMODULE_LIB_DIR}/explicit_layer.d" >> .env
+        fi
+    fi
+
+    echo "DYLD_LIBRARY_PATH=${SUBMODULE_LIB_DIR}" >> .env
+fi
```
Comment (jonjondev) [inline scripts/install.sh:212]: might want to issue a message informing the user that `VULKAN_SDK` was found, and is aborting

File: scripts/install.bat [hunk 13]
```diff
@@ -0,0 +1,197 @@
+@ECHO OFF
```
Comment (jonjondev) [inline scripts/install.bat:1]: all the same concerns raised in the `.sh`

File: README.md [hunk 14]
```diff
@@ -1,152 +1,187 @@
...
-
-#### macOS & Linux
-```bash
-$ make setup
-$ make
-```
+- [Cmake](https://cmake.org/) (all platforms)
+- [GNU Make](https://www.gnu.org/software/make/) (linux & macos)
+- [MingW32-make](https://www.mingw-w64.org/) (windows)
+- [Python](https://www.python.org/) (all platforms)
```
Comment (jonjondev) [inline README.md:41]: what requires Python?

File: README.md [hunk 15]
```diff
@@ -1,152 +1,187 @@
...
-```
+   // Windows
+   > ./scripts/install.bat
+   ```
 
-### Specifying Custom Macro Definitions
-You may also want to pass in your own macro definitions for certain configurations (such as setting log levels). You can pass in your definitions using the `MACRO_DEFS` flag:
+3. This should install all required dependencies. Once completed a `.env` file will be generated with all required variables. If the build is completed with no issue then you can proceed to build the project.
 
+#### Setup without the Vulkan SDK with validation layers
```
Comment (jonjondev) [inline README.md]: I would write this up as an extension or bonus to the previous section

**Review by jonjondev — APPROVED** [review]
Looks good, few comments

File: scripts/mkdir.bat (hunk 4 above)
Comment (Raelr) [reply scripts/mkdir.bat:3]: I actually think we can remove this if we aren't making any directories in the makefile

File: .github/workflows/ubuntu.yml (hunk 5 above)
Comment (Raelr) [reply .github/workflows/ubuntu.yml:22]: These are GUI libraries that are needed by Linux when building the Vulkan Loader.

File: README.md (hunk 14 above)
Comment (Raelr) [reply README.md:41]: The Validation layers has a dependency that uses a Python script. Unfortunately I haven't found a way to get around this limitation just yet.

File: scripts/mkdir.bat (hunk 4 above)
Comment (jonjondev) [reply scripts/mkdir.bat:3]: We will be once merged. Worth dropping until needed?

Comment (Raelr) [reply scripts/mkdir.bat:3]: I don't think so. the MKDIR commands will only work recursively on windows via this script. We do use mkdir when we compile the project. Removing this script breaks the windows compile step in CICD. Strangely enough this isn't an issue on my local machine. 

I say we cop this for now. It isn't a fun solution but it's the best one I can find at the moment.

File: src/main.cpp (hunk 3 above)
Comment (Raelr) [reply src/main.cpp:1]: Yes, but in this case we must define this macro BEFORE we include `volk.h`. There's no way around this issue at the moment.

File: .env.template (hunk 1 above)
Comment (Raelr) [reply .env.template]: Done.

File: scripts/install.sh (hunk 7 above)
Comment (Raelr) [reply scripts/install.sh:38]: No. We don't build `volk.c` but we do need it in order to use `volk.h`.

File: src/main.cpp (hunk 3 above)
Comment (jonjondev) [reply src/main.cpp:1]: Understood. Just checking.

File: scripts/install.sh (hunk 7 above)
Comment (jonjondev) [reply scripts/install.sh:38]: need it in what way?

File: scripts/install.sh (hunk 12 above)
Comment (Raelr) [reply scripts/install.sh:212]: Done.

File: scripts/install.sh (hunk 11 above)
Comment (Raelr) [reply scripts/install.sh]: Yes. I've updated all of the cases where we run a `cd` command

File: scripts/install.sh (hunk 7 above)
Comment (Raelr) [reply scripts/install.sh:38]: If we don't include it the build fails. I'm afraid this isn't just a header only library

File: scripts/install.sh (hunk 8 above)
Comment (Raelr) [reply scripts/install.sh:19]: I'm a little unsure what this means

File: scripts/install.sh (hunk 9 above)
Comment (Raelr) [reply scripts/install.sh]: Done. I've changed most of the instances where we run CD. In the case of the code above, it looks like the script relies on relative paths. I've wrapped the code in parentheses to open a subshell and ensure we don't end up in a random directory.

File: README.md (hunk 15 above)
Comment (Raelr) [reply README.md]: Done.

File: scripts/install.bat (hunk 13 above)
Comment (Raelr) [reply scripts/install.bat:1]: Edited both the `.bat` and `.sh` files. I addressed the concerns you mentioned as best I could. All others have been commented on in their respective sections.

File: scripts/install.sh (hunk 7 above)
Comment (jonjondev) [reply scripts/install.sh:38]: wait, how how does it get included/built??

Comment (jonjondev) [reply scripts/install.sh:38]: are we including `volk.c` somewhere, or are we compiling?

Comment (Raelr) [reply scripts/install.sh:38]: Yes we do. It's one of the includes in our `Makefile` [here](https://github.com/Raelr/vk-sandbox/blob/577b685b63e1855c2fdadc5f193dc55517fb1c70/Makefile#L15)

**Pull request merged**